contracts/buy/PartyBuy.sol

```diff
@@ -0,0 +1,180 @@
+        __ERC721Holder_init();
+        // set storage variables
+        nftContract = IERC721Metadata(_nftContract);
+        tokenId = _tokenId;
+        name = _name;
+        symbol = _symbol;
+        expiresAt = _secondsToTimeout + block.timestamp;
+        require(_maxPrice > 0, "PartyBuy::initialize: must set price higher than 0");
+        maxPrice = _maxPrice;
+        // validate that party split won't retain the total token supply
+        uint256 _remainingBasisPoints = 10000 - TOKEN_FEE_BASIS_POINTS;
+        require(_splitBasisPoints < _remainingBasisPoints, "PartyBuy::initialize: basis points can't take 100%");
+        // validate that a portion of the token supply is not being burned
+        if (_splitRecipient == address(0)) {
+            require(_splitBasisPoints == 0, "PartyBuy::initialize: can't send tokens to burn addr");
+        }
+        splitBasisPoints = _splitBasisPoints;
+        splitRecipient = _splitRecipient;
+        // validate token exists
+        require(_getOwner() != address(0), "PartyBuy::initialize: NFT getOwner failed");
+    }
+
+    // ======== External: Contribute =========
+
+    /**
+     * @notice Contribute to the Party's treasury
+     * while the Party is still active
+     * @dev Emits a Contributed event upon success; callable by anyone
+     */
+    function contribute() external payable nonReentrant {
+        // require that the new total contributed is not greater than
+        // the maximum amount the Party is willing to spend
+        require(totalContributedToParty + msg.value <= getMaximumContributions(), "PartyBuy::contribute: cannot contribute more than max");
+        // continue with shared _contribute flow
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::buy: party not active"
+        );
+        // check that value is not more than the maximum price set at deploy time
+        require(_value <= maxPrice, "PartyBuy::buy: can't spend over max price");
+        // check that value is not more than
+        // the maximum amount the party can spend while paying ETH fee
+        require(_value <= getMaximumSpend(), "PartyBuy::buy: insuffucient funds to buy token plus fee");
+        // execute the calldata on the target contract
+        address(_targetContract).call{value: _value}(_calldata);
+        // NOTE: we don't are if the call succeeded
+        // as long as the NFT is owned by the Party
+        require(_getOwner() == address(this), "PartyBuy::buy: failed to buy token");
+        // set partyStatus to WON
+        partyStatus = PartyStatus.WON;
```

Comment:
wondering if we should do this before `address(_targetContract).call{value: _value}(_calldata);`

worried about a malicious `targetContract` that would expire the party (mark as lost) even though it won

Reply:
hmm okay so a couple things here 
- i am adding a re-entrancy guard to `closeParty` so one should not be able to re-enter either `closeParty` or `buy` - that should've been there in the first place 
- wouldn't doing this *after* the call be safer? if the call somehow set partyStatus to LOST, then we would overwrite it to WON here

Reply:
excellent point about doing it after the call 👍